Content.Server/Physics/Controllers/ChasingWalkSystem.cs

```diff
@@ -0,0 +1,95 @@
+using System.Linq;
+using Content.Server.Physics.Components;
+using Robust.Shared.Random;
+using Robust.Shared.Timing;
+using Robust.Shared.Physics.Systems;
+using Robust.Shared.Physics.Components;
+using Robust.Shared.Map;
+using Robust.Shared.Toolshed.TypeParsers;
+using Robust.Shared.Physics.Controllers;
+
+namespace Content.Server.Physics.Controllers;
+
+/// <summary>
+/// A system which makes its entity chasing another entity with selected component.
+/// </summary>
+public sealed class ChasingWalkSystem : VirtualController
+{
+    [Dependency] private readonly IGameTiming _gameTiming = default!;
+    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly SharedTransformSystem _transform = default!;
+    [Dependency] private readonly EntityLookupSystem _lookup = default!;
+    [Dependency] private readonly SharedPhysicsSystem _physics = default!;
+
+    private EntityQuery<TransformComponent> _xformQuery;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        _xformQuery = GetEntityQuery<TransformComponent>();
+    }
+
+    public override void UpdateBeforeSolve(bool prediction, float frameTime)
+    {
+        base.UpdateBeforeSolve(prediction, frameTime);
+
+        var query = EntityQueryEnumerator<ChasingWalkComponent>();
+        while (query.MoveNext(out var uid, out var chasing))
+        {
+            //Set Velocity to Target
+            if (chasing.NextImpulseTime <= _gameTiming.CurTime)
+            {
+                ForceImpulse(uid, chasing);
+                chasing.NextImpulseTime = _gameTiming.CurTime + TimeSpan.FromSeconds(chasing.ImpulseInterval);
+            }
+            //Change Target
+            if (chasing.NextChangeVectorTime <= _gameTiming.CurTime)
+            {
+                ChangeTarget(uid, chasing);
+
+                var delay = TimeSpan.FromSeconds(_random.NextFloat(chasing.ChangeVectorMinInterval, chasing.ChangeVectorMaxInterval));
+                chasing.NextChangeVectorTime = _gameTiming.CurTime + delay;
+            }
+        }
+    }
+
+    private void ChangeTarget(EntityUid uid, ChasingWalkComponent component)
+    {
+        //We find our coordinates and calculate the radius of the target search.
+        var xform = Transform(uid);
+        var range = component.MaxChaseRadius;
+        var compType = _random.Pick(component.ChasingComponent.Values).Component.GetType();
+        var allEnts = _lookup.GetEntitiesInRange(compType, _transform.GetMapCoordinates(xform), range);
```

Comment:
As above on not re-allocating the set every time, additionally you shouldn't be querying for contained entities.